Kubernetes docs (#953)
lightbend/cloudflow: docs/shared-content-source/docs/modules/develop/pages/cloudflow-configuration-kubernetes.adoc

:page-partial:

include::ROOT:partial$include.adoc[]

== Available Kubernetes Settings 

In this section we are covering the different Kubernetes parameters that are configurable. 

> To set the configuration on a container related to Kubernetes is necessary, first of all, to add that configuration under `cloudflow.streamlets.[streamlet-name].kubernetes`.

Reviewer: ```suggestion
Several Kubernetes settings for a streamlet can be set under `cloudflow.streamlets.[streamlet-name].kubernetes`.
```

Author: nice

> On the followint samples we show configurations under a specific streamlet, `my-streamlet`, that in its complete path is `cloudflow.streamlets.my-streamlet`. It is also possible, as explained in xref:develop:cloudflow-configuration.adoc#_configuring_a_runtime_using_the_runtime_scope[Configuring a Runtime using the runtime Scope], to apply any configuration shown in here to any `runtimes`. The complete path for a runtime is, for instance, `cloudflow.runtimes.akka`.

Reviewer: ```suggestion
On the following samples we show configurations under a specific streamlet, `my-streamlet`, that in its complete path is `cloudflow.streamlets.my-streamlet`. It is also possible, as explained in xref:develop:cloudflow-configuration.adoc#_configuring_a_runtime_using_the_runtime_scope[Configuring a Runtime using the runtime Scope], to apply any configuration shown in here to any `runtimes`. The complete path for a runtime is, for instance, `cloudflow.runtimes.akka`.
```

> On the followint samples we show configurations under a specific streamlet, `my-streamlet`, that in its complete path is `cloudflow.streamlets.my-streamlet`. It is also possible, as explained in xref:develop:cloudflow-configuration.adoc#_configuring_a_runtime_using_the_runtime_scope[Configuring a Runtime using the runtime Scope], to apply any configuration shown in here to any `runtimes`. The complete path for a runtime is, for instance, `cloudflow.runtimes.akka`.

Reviewer: ```suggestion
In the following examples we show configurations for a specific streamlet named `my-streamlet`, under `cloudflow.streamlets.my-streamlet`. It is also possible, as explained in xref:develop:cloudflow-configuration.adoc#_configuring_a_runtime_using_the_runtime_scope[Configuring a Runtime using the runtime Scope], to apply any configuration shown here to a specific `runtime`. For example, Kubernetes settings for all akka streamlets can be set under`cloudflow.runtimes.akka.kubernetes`.
```

> === Resources and Environment Variables

Reviewer: Please create separate sections for configuring resources and configuring environment variables. 
"Configure container resource requests and limits" could be a title for one, "Configure container environment variables" for the other.

> Container resource requirements and environment variables can be set in the `cloudflow.streamlets.[streamlet-name].kubernetes.pods.pod.container.container` section.

Reviewer: ```suggestion
Container resource requirements and environment variables can be set in the `cloudflow.streamlets.[streamlet-name].kubernetes.pods.pod.containers.container` section.
```

The example below shows how resource requirements are set specifically for a `my-streamlet` streamlet:

[source, hocon]
----
cloudflow.streamlets.my-streamlet {
  kubernetes.pods.pod {
    containers.container {
      env = [ 
            { name = "JAVA_OPTS" 
              value = "-XX:MaxRAMPercentage=40.0"
            },{
              name = "FOO"
              value = "BAR"
            }
          ]
      resources {
        requests {
          cpu = "500m"
          memory = "512Mi"
        }
        limits {
          memory = "4Gi"
        }
      }
    }
  }
}
----

The above example shows that `500mcpu` is requested, as well as `512Mi` of memory. 
The pod is limited to use `4Gi` of memory. It also shows how to set the environment variable `JAVA_OPTS=XX:MaxRAMPercentage=40.0` and `FOO=BAR`  

===  Labels

> To set label into a container we need to add the `label` setting as the example below.

Reviewer: ```suggestion
To set container labels, add the `labels` section shown in the example below:
```

[source, hocon]
----
cloudflow.streamlets.my-streamlet {
  kubernetes.pods.pod {
    labels {
      key1 = value1
      key2 = value2
    }
  }
}
----

> The example above shows is adding the labels `key1 = value1` and `key2 = value2` to the pod that contains the streamlet `my-streamlet`. 

Reviewer: ```suggestion
The example above shows how to add the labels `key1 = value1` and `key2 = value2` to the pod for streamlet `my-streamlet`. 
```

===  Annotations

> Similar to labels we can add annotations to any container. This snippet shows how

Reviewer: ```suggestion
Similar to labels we can add annotations to any streamlet pod. The snippet below shows how:
```

[source, hocon]
----
cloudflow {
  streamlets {
    my-streamlet {
      kubernetes {
        pods.pod {
          annotations {
            key1 = annotation1
          }
        }
      }
    }
  }
}
----

> This configuration adds the anotation `key: annotation1` to the container where `my-akka-streamlet` is deployed.

Reviewer: ```suggestion
This configuration adds the annotation `key: annotation1` to the pod for streamlet `my-akka-streamlet`.
```

> Annotations, same as labels, are defined by a key/value and also can be defined specifically for a streamlet or generically to a runtime, like cloudflow.runtime.akka. 

Reviewer: ```suggestion
Annotations, same as labels, are defined by a key/value and also can be defined specifically for a streamlet or for all streamlets of a runtime, under `cloudflow.runtimes.akka.kubernetes.pods.pod.annotations`. 
```

> ===  Ports

Reviewer: ```suggestion
===  Container Ports
```

> To open ports in container we need to add at least the port number. This snippet shows how

Reviewer: ```suggestion
To open ports in containers we need to at least add a port number. The snippet below shows how:
```

[source, hocon]
----
cloudflow {
  streamlets {
    my-streamlet {
      kubernetes {
        pods.pod.containers.container {
          ports = [
            { container-port=9001},
            { container-port=9002, host-ip="10.132.0.123", protocol = "SCTP", host-port=9999, name="user-datagram"}
          ]
        }
      }
    }
  }
}
----

It is possible, as we showed above, to also set host-ip, protocol, host-port and a name. 

> This configuration opens in the container where `my-akka-streamlet` is deployed two ports. One in 9001 as TCP and the other in 9002 as UDP with `Host Ports:9999/UDP`, `name: user-datagram` and `hostIp: 10.132.0.123`

Reviewer: ```suggestion
This configuration adds two container ports to every container where `my-akka-streamlet` is deployed. One in 9001 as TCP and the other in 9002 as UDP with `Host Ports:9999/UDP`, `name: user-datagram` and `hostIp: 10.132.0.123`
```

> For more info on kubernetes API https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#containerport-v1-core[container port]

Reviewer: ```suggestion
For more info see container ports in the Kubernetes API https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#containerport-v1-core[container port]
```

Note:

> Only Akka Streamlets can open ports 

Reviewer: ```suggestion
It is only possible to add container ports to Akka Streamlets, the Flink and Spark operator do not support this.
```

=== Volumes and Volume-Mounts
==== **Mounting Secrets**

> It possible to add `volumes`, that contain a secret, and mount them into the container with `volume-mounts`. Bear in mind that each `kubernetes.pods.pod.containers.container.volume-mounts` name must match a `kubernetes.pods.pod.volumes` name in the config. 

Reviewer: ```suggestion
It possible to add a `volume` that contains a secret, and mount it into the container with `volume-mounts`. Bear in mind that each `kubernetes.pods.pod.containers.container.volume-mounts` name must match a `kubernetes.pods.pod.volumes` name in the config. 
```

          read-only = false
        }
      }
    }
    containers.container {
      volume-mounts {
        foo {
          mount-path = "/mnt/folderA"
          read-only = true
        }
        bar {
          mount-path = "/mnt/folderB"
          read-only = false
        }
      }
    }
  } 
}
----

This configuration will mount an existing Persistent Volume Claim `myclaim1` into each `my-streamlet` container on to the path `/mnt/folderA` as read-only, while `myclaim2` will be mounted on path `/mnt/folderB` as writable. As each `volume-mounts` name must match a `volumes` name, `volume-mounts.foo` matches `volumes.foo` and `volume-mounts.bar` matches `volumes.bar`.

.Configuration available to choose pvc and mount it
[%autowidth]
|===
| resource  | key
| `volumes` | `kubernetes.pods.pod.volumes.[name].pvc.name = [pvc-name]`
| `volumes` | `kubernetes.pods.pod.volumes.[name].pvc.read-only = [true or false]`
| `volume-mounts` | `pod.containers.container.volume-mounts.[name].mount-path = [some-path]`
| `volume-mounts` | `pod.containers.container.volume-mounts.[name].read-only = [true or false]`
|===

=== Pod differences among Akka, Flink and Spark Streamlets.

> Akka runtimes only deploy one pod per streamlet while Flink and Sparks deploy two. Spark creates a Driver pod and a Executor pod, while Flink creates a JobManager and a TaskManager.

Reviewer: ```suggestion
Akka runtimes only deploy one type of pod per streamlet. Spark creates Driver and Executor pods, while Flink creates JobManager and TaskManager pods.
```

> When we are setting Kubernetes configuration via `kubernetes.pods.pod` or `kubernetes.pods.pod.containers.container`, it applies the same settings to all pods. Whether is one or two. 

Reviewer: ```suggestion
When we are setting Kubernetes configuration via `kubernetes.pods.pod` or `kubernetes.pods.pod.containers.container`, it applies the same settings to all types of pod for the runtime used.
```

In the case of Spark we can be more specific by using the following paths:
  
- For the Driver pods: `kubernetes.pods.driver` and `kubernetes.pods.driver.containers.container`     
- For the Executor pods: `kubernetes.pods.executor` and `kubernetes.pods.executor.containers.container`   

> Using the path to the `pod` (driver or executor) or the container it will solely depend on the parameter we are setting. 

Reviewer: ```suggestion
Using the path to the `pod` (driver or executor) settings will only applied to that specific type of pod. 
```

> Flink configuration works similarly to Spark's as we can set the following paths:

Reviewer: ```suggestion
Flink configuration works similarly as we can set the following paths:
```

            mount-path = "/mnt/folderA/mysecret1"
            read-only = true
          }
          baz {
            mount-path = "/mnt/folderB"
            read-only = false
          }
        }
      }
    } 
    executor {
      containers.container {
        volume-mounts {
          foo {
            mount-path = "/mnt/folderA/mysecret1"
            read-only = true
          }
          bar {
            mount-path = "/mnt/folderC/mysecret2"
            read-only = true
          }
        }
      }
    } 
  }
}
---- 

This configuration will mount `mysecret1` into any `driver` pod on path `/mnt/folderA/mysecret1` as read-only and into any `executor. It will mount `myclaim` into any `driver` pod on path `/mnt/folderB` as writable and `mysecret2` will be mounted into any `executor` pod on path `/mnt/folderC/mysecret2`. Inside folders `A` and `C`, you'll find as many files as values are under the https://kubernetes.io/docs/concepts/configuration/secret/#decoding-a-secret[`data`] tag in the Kubernetes Secret mounted. While the mounted `myclaim1` into the `driver` pods will allow to share the contents of folder `/mnt/folderB` among all of them. 

## Exceptions

> Is important to bear in mind that flink configuration will not be able to set specific labels per type of pod (`task-manager` and `job-manager` pods). Only `cloudflow.runtimes.flink.kubernetes.pods.pod.labels` is allowed, cloudflow.runtimes.flink.kubernetes.pods.[pod type].labels is not. It has the same limitation when defining `volumes` and `volume-mounts`. Is only allowed `cloudflow.runtimes.flink.kubernetes.pods.pod.volumes` and `cloudflow.runtimes.flink.kubernetes.pods.pod.containers.container.volume-mounts`.

Reviewer: ```suggestion
Is important to bear in mind that Flink configuration will not be able to set specific labels per type of pod (`task-manager` and `job-manager` pods). Only `cloudflow.runtimes.flink.kubernetes.pods.pod.labels` is allowed, cloudflow.runtimes.flink.kubernetes.pods.[pod type].labels is not. It has the same limitation when defining `volumes` and `volume-mounts`. Is only allowed `cloudflow.runtimes.flink.kubernetes.pods.pod.volumes` and `cloudflow.runtimes.flink.kubernetes.pods.pod.containers.container.volume-mounts`.
```

:page-partial:

include::ROOT:partial$include.adoc[]

== Available Kubernetes Settings 

> In this section we are covering the different Kubernetes settings that are configurable. 

Reviewer: ```suggestion
In this section we are covering the Kubernetes settings that you can configure in Cloudflow. 
```

> In this section we are covering the different Kubernetes parameters that are configurable. 

Reviewer: ```suggestion
In this section we are covering the different Kubernetes settings that are configurable. 
```

Several Kubernetes settings for a streamlet can be set under `cloudflow.streamlets.[streamlet-name].kubernetes`.

> On the following samples we show configurations under a specific streamlet, `my-streamlet`, its complete path is `cloudflow.streamlets.my-streamlet`. It is also possible, as explained in xref:develop:cloudflow-configuration.adoc#_configuring_a_runtime_using_the_runtime_scope[Configuring a Runtime using the runtime Scope], to apply any configuration shown in here to any `runtimes`. The complete path for a runtime is, for instance, `cloudflow.runtimes.akka`.

Reviewer: ```suggestion
In the following examples we show configurations under a specific streamlet, `my-streamlet`, its complete path is `cloudflow.streamlets.my-streamlet`. It is also possible, as explained in xref:develop:cloudflow-configuration.adoc#_configuring_a_runtime_using_the_runtime_scope[Configuring a Runtime using the runtime Scope], to apply any configuration shown in here to any `runtimes`. The complete path for a runtime is, for instance, `cloudflow.runtimes.akka`.
```

> On the following samples we show configurations under a specific streamlet, `my-streamlet`, that in its complete path is `cloudflow.streamlets.my-streamlet`. It is also possible, as explained in xref:develop:cloudflow-configuration.adoc#_configuring_a_runtime_using_the_runtime_scope[Configuring a Runtime using the runtime Scope], to apply any configuration shown in here to any `runtimes`. The complete path for a runtime is, for instance, `cloudflow.runtimes.akka`.

Reviewer: ```suggestion
On the following samples we show configurations under a specific streamlet, `my-streamlet`, its complete path is `cloudflow.streamlets.my-streamlet`. It is also possible, as explained in xref:develop:cloudflow-configuration.adoc#_configuring_a_runtime_using_the_runtime_scope[Configuring a Runtime using the runtime Scope], to apply any configuration shown in here to any `runtimes`. The complete path for a runtime is, for instance, `cloudflow.runtimes.akka`.
```

> === Resources

Reviewer: ```suggestion
=== Resource Requirements
```

Container resource requirements can be set in the `cloudflow.streamlets.[streamlet-name].kubernetes.pods.pod.containers.container` section.
The example below shows how resource requirements are set specifically for a `my-streamlet` streamlet:

[source, hocon]
----
cloudflow.streamlets.my-streamlet {
  kubernetes.pods.pod {
    containers.container {
      resources {
        requests {
          cpu = "500m"
          memory = "512Mi"
        }
        limits {
          memory = "4Gi"
        }
      }
    }
  }
}
----

> The above example shows that `500mcpu` is requested, as well as `512Mi` of memory while being limited to use `4Gi` of memory.   

Reviewer: ```suggestion
The above example shows that `500mcpu` and `512Mi` of memory is requested, while the memory limit is set to `4Gi`.   
```

=== Environment Variables

Container environment variables can be set in the `cloudflow.streamlets.[streamlet-name].kubernetes.pods.pod.containers.container` section.
The example below shows how resource requirements are set specifically for a `my-streamlet` streamlet:

[source, hocon]
----
cloudflow.streamlets.my-streamlet {
  kubernetes.pods.pod {
    containers.container {
      env = [ 
            { name = "JAVA_OPTS" 
              value = "-XX:MaxRAMPercentage=40.0"
            },{
              name = "FOO"
              value = "BAR"
            }
          ]
    }
  }
}
----

> The above example shows how to set the environment variable `JAVA_OPTS=XX:MaxRAMPercentage=40.0` and `FOO=BAR` 

Reviewer: ```suggestion
The above example shows how to set the environment variables `JAVA_OPTS=XX:MaxRAMPercentage=40.0` and `FOO=BAR` 
```

===  Labels

To set pod labels, add the `labels` section shown in the example below:

[source, hocon]
----
cloudflow.streamlets.my-streamlet {
  kubernetes.pods.pod {
    labels {
      key1 = value1
      key2 = value2
    }
  }
}
----

The example above shows how to add the labels `key1 = value1` and `key2 = value2` to the pod for streamlet `my-streamlet`. 

===  Annotations

Similar to labels we can add annotations to any streamlet pod. The snippet below shows how:

[source, hocon]
----
cloudflow {
  streamlets {
    my-streamlet {
      kubernetes {
        pods.pod {
          annotations {
            key1 = annotation1
          }
        }
      }
    }
  }
}
----

This configuration adds the annotation `key: annotation1` to the pod for streamlet `my-akka-streamlet`.

> Annotations, same as labels, are defined by a key/value and also can be defined specifically for a streamlet or for all streamlets of a runtime, under `cloudflow.runtimes.akka.kubernetes.pods.pod.annotations`. 

Reviewer: ```suggestion
Annotations, similar to labels, are defined by a key/value and also can be defined specifically for a streamlet or for all streamlets of a runtime, under `cloudflow.runtimes.akka.kubernetes.pods.pod.annotations`. 
```
